README.md
# Table of Contents

* [Getting Started](#getting-started)
* [Installation](#installation)
* [Basic usage](#basic-usage)
* [Advance usage](#advance-usage)
* [License](#license)
* [Contributing](#contributing)
* [FAQ](#faq)

# Getting Started

This software tool is designed to help with version control of software projects. It is designed to work with git repositories and it is written in Python.

It uses a combination of git tags and git commits to calculate the current version of the software. It also allows you to generate a changelog file based on git commits.

Version calculation is based on:

* [SymVer 2.0.0](https://semver.org/)
* [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)

The default keywords used to calculate the version are:

* `break` - for breaking changes
* `feat` - for new features
* `fix` - for bug fixes

All other keywords are treated as non changing the version, but they are forming a release candidate (`.rcX`).

# Installation

This software requires the following packages to be installed:

* `python3`
* `python3-pip`
* `pyinstaller`

You can install these packages on Ubuntu with the following command:

```bash
sudo make configure
```

After satysfying the above requirements, you can build the executable of this software with the following command:

```bash
make build
```

To finish installation, all necessary files mute be copied to coresponding directories. This can be done with the following command:

```bash
sudo make install
```

This script does the following:

1. Copies the `version-control` executable to `/usr/local/bin`
2. Copies bash autocompletion script to `/etc/bash_completion.d`
3. Creates a shortcut symbolic link named `vc` in `/usr/local/bin` to the main executable

# Basic usage

Typical usage of this software is as follows:

1. Calculate and show current software version in current repository:

    ```bash
    version-control show
    ```

1. Find specyfic version in repository that coresponds to `1.0.0`:

    ```bash
    version-control find --version 1.0.0
    ```

2. Find all versions of the software in specyfic repository:

    ```bash
    version-control find --path /home/user/Projects/MyProject
    ```

3. Find all version of the software in current repository, filter by author and show result is json format:

    ```bash
    version-control find --author "John Doe" --format json
    ```

4. Find all versions of the software in current repository and change output schema to `version` + `date` pair:

    ```bash
    version-control find --schema version date
    ```

5. Generate a changelog file and save it to specified location:

    ```bash
    version-control generate --output docs/CHANGELOG.md
    ```

6. Sign a file with current repository version (i.e rename a file to `file-version`):

    ```bash
    version-control sign --file file
    ```

# Advance usage

This software was build to support CI/CD procedure for standalone project. You can use it to automate versioning and changelog generation.

For example, you can use it in your `Makefile` to automate versioning and changelog generation by adding this target:

```makefile
release:
    git add . && git commit # Commit all changes to update version
    git push
    g++ -o myapp main.cpp # Build your C/C++ application
    version-control generate --output CHANGELOG.md # Generate changelog
    version-control find --schema version date > VERSIONS.txt # Save version history
    version-control sign --file myapp # Sign your application with version number ie. myapp-0.4.20
```

# License

This software is licensed under the MIT license. See the [LICENSE](LICENSE) file for more information.

# Contributing

This software is open source and contributions are welcome. See the [CONTRIBUTING](CONTRIBUTING.md) file for more information.

# FAQ

1. **Question:** Why my changelog contains strange strings in places related to Github?

    > Make sure that you repository has a remote named `origin` and that you have access to it. This software uses `git log` command to generate changelog and it uses `origin` remote to generate links to commits.

2. **Question:** Why some of my changelog links are not working properly ?

    > Make sure you have all of your commits pushed to remote repository (for example with `git push` command) in order to make hyperlinks work properly.

3. **Question:** Why there are some strange unrendered characters in my changelog ?

    > `version-control` software uses basic inline `HTML` tags to format changelog. Make sure that your markdown renderer supports inline `HTML` tags. It is recommended to use one of VSCode extensions to render markdown files.

4. **Question:** Does `find` module support asterisk wildcards ?

    > Yes, you can use asterisk wildcards in `find` module. For example, you can use `version-control find --author "John*"` to find all versions created by authors with name starting with `John`.

CHANGELOG.md

<style>
   .gone {
      color: #CD5C5C;
      font-size: 16px;
      font-weight: 360;
   }
   .new {
      color: #2E8B57;
      font-size: 16px;
      font-weight: 380;
   }
   .modified {
      color: #ffc87c;
      font-size: 16px;
      font-weight: 350;
   }
</style>

# version-control-v0.25.2 (by ErnestSzczepaniak)

This document was automatically generated by **version-control** software tool, created by Ernest Szczepaniak @ CODWAY, 2023.
# Table of contents

* [Overview](#overview)
  * [Remote address](#remote-address)
  * [Connection protocol](#connection-protocol)
  * [Branches](#branches)
  * [Current version](#current-version)
  * [Contributors](#contributors)
* [Statistics](#statistics)
  * [Project timeframe](#project-timeframe)
  * [Code frequency](#code-frequency)
  * [Commit structure](#commit-structure)
* [Version history](#version-history)
  * [Newest version](#newest-version)
  * [Oldest version](#oldest-version)
  * [Version table](#version-table)
* [Changelog](#changelog)
# Overview

#### Remote address:
* https://github.com/ErnestSzczepaniak/version-control

#### Connection protocol:
* ssh

#### Branches:
* main [current] (52 commits)

#### Current version:
* 0.25.2

#### Contributors:
* Ernest Szczepaniak (ernest.szczepaniak.en2@gmail.com)

# Statistics

#### Project timeframe:
* 20.04.2023 - 12.05.2023 (23 days)

#### Code frequency:
* 51 commits (2.22 / day)
* 150 files changed (6.52 / day)
* 2183 insertions (94.91 / day)
* 1839 deletions (79.96 / day)

#### Commit structure:
* feat - 25 (49.02%)
* fix - 20 (39.22%)
* style - 6 (11.76%)

# Version history

#### Newest version:
* [0.25.2](#0252)

#### Oldest version:
* [0.0.1](#001)

#### Version table based on new features:

<div style="margin-left: 30px;">

| New features | Fixes | | | | | | |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| include styling in changelog file and add colored files modyfications |  [0.25.2](#0252) | [0.25.1](#0251) | [0.25.0](#0250) |
| add double star filtering in find module |  [0.24.0](#0240) |
| add (new) and (gone) events in changes list |  [0.23.1](#0231) | [0.23.0](#0230) |
| add Color flag to find module |  [0.22.0](#0220) |
| add advanced usage part in README.md file |  [0.21.0](#0210) |
| insert basic text to README.md file |  [0.20.0](#0200) |
| add css div styling to each version paragraph |  [0.19.0](#0190) |
| calculate commit duration and rework CHANGELOG apperance |  [0.18.0](#0180) |
| add vc symbolic link for faster command typing |  [0.17.0](#0170) |
| include contributor email address |  [0.16.2](#0162) | [0.16.1](#0161) | [0.16.0](#0160) |
| add parent hash for each commit |  [0.15.0](#0150) |
| add basic auto-completion |  [0.14.1](#0141) | [0.14.0](#0140) |
| add version history table with feature and all fixes |  [0.13.1](#0131) | [0.13.0](#0130) |
| add quick description that changelog was generated automatically |  [0.12.0](#0120) |
| change Url to dataclass, add protocol in Overview |  [0.11.0](#0110) |
| add arguments.py file to store all arguments in one and accessible form |  [0.10.0](#0100) |
| add sub links in Table of contents, add additional regex patterns, combine github functions |  [0.9.4.rc3](#094rc3) | [0.9.4.rc2](#094rc2) | [0.9.4.rc1](#094rc1) | [0.9.3](#093) | [0.9.2](#092) | [0.9.1](#091) | [0.9.0](#090) |
| add regex finding in log and show commands |  [0.8.0](#080) |
| add edit number to each file in files changed section |  [0.7.3](#073) | [0.7.2](#072) | [0.7.1](#071) | [0.7.0](#070) |
| add files changed tracker with reworked insertions. |  [0.6.0](#060) |
| split changelog to sections |  [0.5.0](#050) |
| add release candidates numbering |  [0.4.1](#041) | [0.4.0.rc2](#040rc2) | [0.4.0.rc1](#040rc1) |
| add branches in changelog |  [0.3.0.rc3](#030rc3) | [0.3.0.rc2](#030rc2) | [0.3.0.rc1](#030rc1) |
| check commit spacing |  [0.2.2](#022) | [0.2.1](#021) | [0.2.0.rc2](#020rc2) | [0.2.0.rc1](#020rc1) |
| test 2 line body |  [0.1.2](#012) | [0.1.1](#011) | [0.1.0](#010) |
| remove empty spaces |  [0.0.1](#001) |
</div>

# Changelog

## [0.25.2](https://github.com/ErnestSzczepaniak/version-control/tree/9a613e3ac927c7dcd91366be1495c82f5034fb0f)

<div style="margin-left: 30px; margin-bottom: 30px;">

&#9737; &nbsp; &nbsp; [12.05.2023] [[bcca541](https://github.com/ErnestSzczepaniak/version-control/commit/bcca541) &#8674; [9a613e3](https://github.com/ErnestSzczepaniak/version-control/commit/9a613e3)] (fix) - add additional margin to commit body section in each paragraph (Ernest Szczepaniak @ 13:59:14)

&#9998; &nbsp; &nbsp; 4 files changed (24 insertions, 10 deletions):

   * <div class="modified"> README.md [+2, -1] </div>
   * <div class="modified"> find.py [+5, -1] </div>
   * <div class="modified"> generate.py [+11, -2] </div>
   * <div class="modified"> github.py [+6, -6] </div>

&#9776; &nbsp; &nbsp; Detailed description:

   * Commit hash - `9a613e3ac927c7dcd91366be1495c82f5034fb0f`
   * Parent hash - `bcca541a18099070137ce12d797d248adb13dc42`
   * Duration - `0:41:06`

</div>

## [0.25.1](https://github.com/ErnestSzczepaniak/version-control/tree/bcca541a18099070137ce12d797d248adb13dc42)

<div style="margin-left: 30px; margin-bottom: 30px;">

&#9737; &nbsp; &nbsp; [12.05.2023] [[7a19581](https://github.com/ErnestSzczepaniak/version-control/commit/7a19581) &#8674; [bcca541](https://github.com/ErnestSzczepaniak/version-control/commit/bcca541)] (fix) - add more FAQ questions (Ernest Szczepaniak @ 13:18:08)

&#9998; &nbsp; &nbsp; 1 file changed (7 insertions, 3 deletions):

   * <div class="modified"> README.md [+7, -3] </div>

&#9776; &nbsp; &nbsp; Detailed description:

   * Commit hash - `bcca541a18099070137ce12d797d248adb13dc42`
   * Parent hash - `7a19581c7edbbe0791e26867f49dea45803e47bb`
   * Duration - `0:09:54`

</div>

## [0.25.0](https://github.com/ErnestSzczepaniak/version-control/tree/7a19581c7edbbe0791e26867f49dea45803e47bb)

<div style="margin-left: 30px; margin-bottom: 30px;">


generate.py
import github, mark, datetime, pathlib, arguments
from typing import List
from github import Commit, Branch

ARGUMENTS = [
    arguments.PATH,
    arguments.MAJOR,
    arguments.MINOR,
    arguments.PATCH,
    arguments.OUTPUT
]

style = """
<style>
   .gone {
      color: #CD5C5C;
      font-size: 16px;
      font-weight: 360;
   }
   .new {
      color: #2E8B57;
      font-size: 16px;
      font-weight: 380;
   }
   .modified {
      color: #ffc87c;
      font-size: 16px;
      font-weight: 350;
   }
</style>
"""

def add_remote_address(md: mark.Markdown, url: github.Url):
    md.h4('Remote address:')
    md.item(f'https://github.com/{url.username}/{url.repository}')
    md.text('')

def add_connection_protocol(md: mark.Markdown, url: github.Url):
    md.h4('Connection protocol:')
    md.item(f'{url.protocol}')
    md.text('')

def add_branch_list(md: mark.Markdown, branches: List[Branch]):
    md.h4('Branches:')
    for branch in branches:
        if branch.active:
            md.item(f'{branch.name} [current] ({branch.commits} commits)')
        else:
            md.item(f'{branch.name} ({branch.commits} commits)')
    md.text('')


def add_current_version(md: mark.Markdown, commits: List[Commit]):
    md.h4('Current version:')
    md.item(f'{commits[0].version}')
    md.text('')

def add_contributors(md: mark.Markdown, commits: List[Commit]):
    md.h4('Contributors:')
    contributors = {}
    for commit in commits:
        if commit.author not in contributors:
            contributors[commit.author] = commit.email

    for contributor in contributors:
        md.item(f'{contributor} ({contributors[contributor]})')

    md.text('')

def add_project_timeframe(md: mark.Markdown, commits: List[Commit]):
    md.h4('Project timeframe:')
    date_start = commits[-1].date
    date_end = commits[0].date
    period = datetime.datetime.strptime(date_end, '%d.%m.%Y') - datetime.datetime.strptime(date_start, '%d.%m.%Y')
    days = period.days + 1
    md.item(f'{date_start} - {date_end} ({days} days)')
    md.text('')

def add_code_frequency(md: mark.Markdown, commits: List[Commit]):
    md.h4('Code frequency:')
    date_start = commits[-1].date
    date_end = commits[0].date
    period = datetime.datetime.strptime(date_end, '%d.%m.%Y') - datetime.datetime.strptime(date_start, '%d.%m.%Y')
    period = period.days + 1
    md.item(f'{len(commits)} commits ({round(len(commits) / period, 2)} / day)')

    total_files_changed = 0
    total_insertions = 0
    total_deletions = 0

    for commit in commits:
        total_files_changed += len(commit.changes)
        total_insertions += sum(change.insertions for change in commit.changes)
        total_deletions += sum(change.deletions for change in commit.changes)

    md.item(f'{total_files_changed} files changed ({round(total_files_changed / period, 2)} / day)')
    md.item(f'{total_insertions} insertions ({round(total_insertions / period, 2)} / day)')
    md.item(f'{total_deletions} deletions ({round(total_deletions / period, 2)} / day)')
    md.text('')

def add_commit_structure(md: mark.Markdown, commits: List[Commit]):
    md.h4('Commit structure:')
    occurences = {}
    for commit in commits:
        if commit.keyword not in occurences:
            occurences[commit.keyword] = 0
        occurences[commit.keyword] += 1
    occurences = {k: v for k, v in sorted(occurences.items(), key=lambda item: item[1], reverse=True)}
    for keyword in occurences:
        percentage = round(occurences[keyword] / len(commits) * 100, 2)
        md.item(f'{keyword} - {occurences[keyword]} ({percentage}%)')
    md.text('')

def add_version_oldest(md: mark.Markdown, commits: List[Commit]):
    md.h4('Oldest version:')
    link = commits[-1].version.replace('.', '')
    md.item(f'[{commits[-1].version}](#{link})')
    md.text('')

def add_version_newest(md: mark.Markdown, commits: List[Commit]):
    md.h4('Newest version:')
    link = commits[0].version.replace('.', '')
    md.item(f'[{commits[0].version}](#{link})')
    md.text('')

def add_version_table(md: mark.Markdown, commits: List[Commit]):
    md.h4('Version table based on new features:')
    md.text('')
    md.text('<div style="margin-left: 30px;">')
    md.text('')

    groups = {}

    for commit in commits:
        version = commit.version
        major, minor = version.split('.')[:2]
        header = f'{major}.{minor}'
        if header not in groups:
            groups[header] = []
        groups[header].append(commit)

    max_items = max(len(groups[key]) for key in groups)

    table_rows = '| New features | Fixes | ' + '| ' * (max_items - 2 ) + '|'

    cendating_rows = '| :- ' + '| :-: ' * max_items + '|'

    md.text(table_rows)
    md.text(cendating_rows)

    for key in groups:
        row = f'| {groups[key][-1].subject} | '
        for commit in groups[key]:
            link = commit.version.replace('.', '')
            row += f' [{commit.version}](#{link}) |'
        md.text(row)

    md.text('</div>')
    md.text('')

def execute(**kwargs):

    path_repository = pathlib.Path(kwargs['path']).absolute().as_posix()

    client = github.Client(path_repository)

    url = client.url()

    commits = client.commits(kwargs['major'], kwargs['minor'], kwargs['patch'])

    if commits is None: return
    
    commits.reverse()

    branches = client.branches()

    md = mark.Markdown()

    md.text(style)

    md.h1(f'{url.repository}-v{commits[0].version} (by {url.username})')
    md.text('')

    md.text('This document was automatically generated by **version-control** software tool, created by Ernest Szczepaniak @ CODWAY, 2023.')

    md.h1('Table of contents')
    md.text('')

    md.item('[Overview](#overview)')
    md.text('  * [Remote address](#remote-address)')
    md.text('  * [Connection protocol](#connection-protocol)')
    md.text('  * [Branches](#branches)')
    md.text('  * [Current version](#current-version)')
    md.text('  * [Contributors](#contributors)')

    md.item('[Statistics](#statistics)')
    md.text('  * [Project timeframe](#project-timeframe)')
    md.text('  * [Code frequency](#code-frequency)')
    md.text('  * [Commit structure](#commit-structure)')

    md.item('[Version history](#version-history)')
    md.text('  * [Newest version](#newest-version)')
    md.text('  * [Oldest version](#oldest-version)')
    md.text('  * [Version table](#version-table)')
    
    md.item('[Changelog](#changelog)')

    md.h1('Overview')
    md.text('')

    add_remote_address(md, url)
    add_connection_protocol(md, url)
    add_branch_list(md, branches)
    add_current_version(md, commits)
    add_contributors(md, commits)

    md.h1('Statistics')
    md.text('')

    add_project_timeframe(md, commits)
    add_code_frequency(md, commits)
    add_commit_structure(md, commits)

    md.h1('Version history')
    md.text('')

    add_version_newest(md, commits)
    add_version_oldest(md, commits)
    add_version_table(md, commits)

    md.h1('Changelog')
    md.text('')

    last_version = None

    for commit in commits:

        if last_version != commit.version:

            link_tree = f'https://github.com/{url.username}/{url.repository}/tree/{commit.hash}'

            md.h2(f'[{commit.version}]({link_tree})')
            md.text('')
            md.text('<div style="margin-left: 30px; margin-bottom: 30px;">')

        link_commit = f'https://github.com/{url.username}/{url.repository}/commit/{commit.hash[0:7]}'
        link_parent = f'https://github.com/{url.username}/{url.repository}/commit/{commit.parent[0:7]}'

        md.text('')

        md.text(f'&#9737; &nbsp; &nbsp; [{commit.date}] [[{commit.parent[0:7]}]({link_parent}) &#8674; [{commit.hash[0:7]}]({link_commit})] ({commit.keyword}) - {commit.subject} ({commit.author} @ {commit.time})')

        if len(commit.body) > 1:
            md.text('')
            md.text('<div style="margin-left: 29px;">')
            md.text('')

            md.text('   ```')
            for element in commit.body[:-1]:
                md.text(f'{element}')
            md.text('   ```')

            md.text('</div>')
            md.text('')

        files_changed = len(commit.changes)
        insertions = sum(change.insertions for change in commit.changes)
        deletions = sum(change.deletions for change in commit.changes)

        md.text('')
        if files_changed == 1:
            md.text(f'&#9998; &nbsp; &nbsp; {files_changed} file changed ({insertions} insertions, {deletions} deletions):')
        else:
            md.text(f'&#9998; &nbsp; &nbsp; {files_changed} files changed ({insertions} insertions, {deletions} deletions):')
        md.text('')

        for change in commit.changes:
            if change.event == '':
                md.text(f'   * <div class="modified"> {change.filename} [+{change.insertions}, -{change.deletions}] </div>')
            elif change.event == 'new':
                md.text(f'   * <div class="new"> {change.filename} [+{change.insertions}, -{change.deletions}] </div>')
            elif change.event == 'gone':
                md.text(f'   * <div class="gone"> {change.filename} [+{change.insertions}, -{change.deletions}] </div>')


        md.text('')

        md.text(f'&#9776; &nbsp; &nbsp; Detailed description:')
        md.text('')

        md.text(f'   * Commit hash - `{commit.hash}`')
        md.text(f'   * Parent hash - `{commit.parent}`')
        md.text(f'   * Duration - `{commit.duration}`')

        md.text('')
        md.text('</div>')
        md.text('')

        last_version = commit.version

    if kwargs['output'] == None:
        print(md.string)
        return

    with open(kwargs['output'], 'w') as file:
        file.write(md.string)

